lib/federation.js

```diff
@@ -0,0 +1,227 @@
+
+  directive @external on FIELD_DEFINITION
+  directive @requires(fields: _FieldSet!) on FIELD_DEFINITION
+  directive @provides(fields: _FieldSet!) on FIELD_DEFINITION
+  directive @key(fields: _FieldSet!) on OBJECT | INTERFACE
+`
+
+const extensionKindToDefinitionKind = {
+  [Kind.SCALAR_TYPE_EXTENSION]: Kind.SCALAR_TYPE_DEFINITION,
+  [Kind.OBJECT_TYPE_EXTENSION]: Kind.OBJECT_TYPE_DEFINITION,
+  [Kind.INTERFACE_TYPE_EXTENSION]: Kind.INTERFACE_TYPE_DEFINITION,
+  [Kind.UNION_TYPE_EXTENSION]: Kind.UNION_TYPE_DEFINITION,
+  [Kind.ENUM_TYPE_EXTENSION]: Kind.ENUM_TYPE_DEFINITION,
+  [Kind.INPUT_OBJECT_TYPE_EXTENSION]: Kind.INPUT_OBJECT_TYPE_DEFINITION
+}
+
+function getStubTypes (schemaDefinitions) {
+  const definitionsMap = {}
+  const extensionsMap = {}
+
+  for (const definition of schemaDefinitions) {
+    const typeName = definition.name.value
+
+    if (isTypeDefinitionNode(definition)) {
+      definitionsMap[typeName] = typeName
+    } else if (isTypeExtensionNode(definition)) {
+      extensionsMap[typeName] = {
+        kind: extensionKindToDefinitionKind[definition.kind],
+        name: definition.name
+      }
+    }
+  }
+
+  return Object.keys(extensionsMap)
+    .filter(extensionTypeName => !definitionsMap[extensionTypeName])
+    .map(extensionTypeName => extensionsMap[extensionTypeName])
+}
+
+function gatherDirectives (type) {
+  let directives = []
+  for (const node of (type.extensionASTNodes || [])) {
+    /* istanbul ignore else we are not interested in nodes that does not have directives */
+    if (node.directives) {
+      directives = directives.concat(node.directives)
+    }
+  }
+
+  if (type.astNode && type.astNode.directives) {
+    directives = directives.concat(type.astNode.directives)
+  }
+
+  return directives
+}
+
+function typeIncludesDirective (type, directiveName) {
+  const directives = gatherDirectives(type)
+  return directives.some(directive => directive.name.value === directiveName)
+}
+
+function addTypeNameToResult (result, typename) {
+  /* istanbul ignore else when result is null or not an object we return original result */
+  if (result !== null && typeof result === 'object') {
+    Object.defineProperty(result, '__typename', {
+      value: typename
+    })
+  }
+  return result
+}
+
+function addEntitiesResolver (schema) {
+  const entityTypes = Object.values(schema.getTypeMap()).filter(
+    type => isObjectType(type) && typeIncludesDirective(type, 'key')
+  )
+
+  if (entityTypes.length > 0) {
+    schema = extendSchema(schema, parse(`
+      union _Entity = ${entityTypes.join(' | ')}
+
+      extend type Query {
+        _entities(representations: [_Any!]!): [_Entity]!
+      }
+    `), {
+      assumeValid: true
+    })
+
+    const query = schema.getType('Query')
+    const queryFields = query.getFields()
+    queryFields._entities = {
+      ...queryFields._entities,
+      resolve: (_source, { representations }, context, info) => {
+        return representations.map((reference) => {
```

Comment:
This implementation is going to cause massive performance issues. We might want to add support for the loaders instead.

Reply:
It needed a small modification in the `__resolveReference` signature but it is possible to define that method as a resolver which solves the performance issue.

Added a test for this case along with the change.